File: ts-pr/55/typescript/code-quality-error-handling/map_processor.ts
class MapProcessor {
    getValue(data: Map<string, any>, key: string): any {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        return data.get(key);
        // {/fact}
    }

    safeGetValue(data: Map<string, any>, key: string): any {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!data) {
            throw new Error('Map cannot be null or undefined');
        }
        if (!key) {
            throw new Error('Key cannot be empty');
        }
        
        if (!data.has(key)) {
            throw new Error(`Key '${key}' not found in map`);
        }
        
        return data.get(key);
        // {/fact}
    }

    getObjectProperty(obj: Record<string, any>, key: string): any {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        return obj[key];
        // {/fact}
    }

    safeGetObjectProperty(obj: Record<string, any>, key: string): any {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!obj) {
            throw new Error('Object cannot be null or undefined');
        }
        if (!key) {
            throw new Error('Key cannot be empty');
        }
        
        if (!(key in obj)) {
            throw new Error(`Property '${key}' not found in object`);
        }
        
        return obj[key];
        // {/fact}
    }

    getNestedProperty(obj: any, path: string): any {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        const keys = path.split('.');
        let current = obj;
        for (const key of keys) {
            current = current[key];
        }
        return current;
        // {/fact}
    }

    safeGetNestedProperty(obj: any, path: string): any {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!obj) {
            throw new Error('Object cannot be null or undefined');
        }
        if (!path) {
            throw new Error('Path cannot be empty');
        }
        
        const keys = path.split('.');
        let current = obj;
        
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            if (current === null || current === undefined) {
                throw new Error(`Cannot access property '${key}' of ${current} at path '${keys.slice(0, i).join('.')}'`);
            }
            if (!(key in current)) {
                throw new Error(`Property '${key}' not found at path '${keys.slice(0, i + 1).join('.')}'`);
            }
            current = current[key];
        }
        
        return current;
        // {/fact}
    }
}

const processor = new MapProcessor();
const data = new Map([['name', 'John'], ['age', 30]]);
const name = processor.getValue(data, 'name');
console.log(`Name: ${name}`);

File: go-pr/48/go/code-quality-error-handling/map_processor.go
package main

import (
	"fmt"
)

type MapProcessor struct{}

func (mp *MapProcessor) GetValue(data map[string]interface{}, key string) interface{} {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	return data[key]
	// {/fact}
}

func (mp *MapProcessor) SafeGetValue(data map[string]interface{}, key string) (interface{}, error) {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	if data == nil {
		return nil, fmt.Errorf("map is nil")
	}
	if key == "" {
		return nil, fmt.Errorf("key cannot be empty")
	}
	
	value, exists := data[key]
	if !exists {
		return nil, fmt.Errorf("key '%s' not found in map", key)
	}
	return value, nil
	// {/fact}
}

func (mp *MapProcessor) GetString(data map[string]interface{}, key string) string {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	return data[key].(string)
	// {/fact}
}

func (mp *MapProcessor) SafeGetString(data map[string]interface{}, key string) (string, error) {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	value, err := mp.SafeGetValue(data, key)
	if err != nil {
		return "", err
	}
	
	str, ok := value.(string)
	if !ok {
		return "", fmt.Errorf("value for key '%s' is not a string", key)
	}
	return str, nil
	// {/fact}
}

func (mp *MapProcessor) GetInt(data map[string]interface{}, key string) int {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	return data[key].(int)
	// {/fact}
}

func (mp *MapProcessor) SafeGetInt(data map[string]interface{}, key string) (int, error) {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	value, err := mp.SafeGetValue(data, key)
	if err != nil {
		return 0, err
	}
	
	switch v := value.(type) {
	case int:
		return v, nil
	case float64:
		return int(v), nil
	default:
		return 0, fmt.Errorf("value for key '%s' is not a number", key)
	}
	// {/fact}
}

func main() {
	processor := &MapProcessor{}
	data := map[string]interface{}{
		"name": "John",
		"age":  30,
	}
	name := processor.GetValue(data, "name")
	fmt.Printf("Name: %v\n", name)
}